src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -745,6 +747,73 @@ class StringGenerator : public Generator<std::string> {
     const size_t _alphabetLength;
 };
 
+/** `{^ChooseFromDataset:{...}` */
+class RandomDataFromDataset : public Generator<std::string> {
+
+public:
+    RandomDataFromDataset(const Node& node, GeneratorArgs generatorArgs)
+        : _rng{generatorArgs.rng},
+          _id{generatorArgs.actorId},
+          _path{node["path"].maybe<std::string>().value()},
+          _pathLength{_path.size()}
+      {
+
+
+        if (_pathLength <= 0) {
+            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+            "RandomDataFromDataset requieres non-empty path"));
+            }
+
+            /* To avoid reading and storing the datasets multiple times
+               we check if the path already exists in the map (as key). If
+               it doesn't exist, we open the file and store its contents. 
+               Section locked with a mutex */
+            
+            _dataset_mutex.lock();
```

Comment:
+1 on std::lock_guard. 

Do we need to lock the the mutex before checking the _dataset.count? It seems like we could do an early exit if (_dataset.count(_path) > 0). Obviously if false, we need to recheck after getting the lock. This is likely a bad idea unless we end up stalling a lot of threads here.

Reply:
FWIW, unordered_map::count is a const function and therefore threadsafe for the container. So is the boolean function contains, which might be a better fit for the condition below. 

We should be able to safely insert into two different entries in the unordered map at the same time from different threads, so the locking could be done by path name. 
Again, I'm likely over thinking this.

Reply:
I think checking the count makes sense; it'd be best to not queue everyone up waiting on the lock to check if the data is present if it actually is present. IIRC from the slow startup times ticket, evaluating generators all this was a nontrivial part of initialization and serializing it too much would make it far worse.

Though I imagine if we have 100 threads in an actor configured similarly, they'll all end up queuing up here at about the same time?

Reply:
- Modified the code so it uses std::lock_guard instead of locking/unlocking.
- Moved the load to its own method.
- Now, during evaluate() call there is a check to see if the path is present in memory. If so we just continue, if not, we call loadDataset method protected by a mutex.

It is true that when we start the load the 100 threads will queue about the same time while one is loading the data, but with the current code it will only happen once per dataset loaded. Then, once it is in memory, the mutex and the loadDataset() is not called anymore. I think this is the best approach here.